src/Stratis.Bitcoin.Features.Consensus/Rules/CommonRules/CoinviewRule.cs

```diff
@@ -0,0 +1,408 @@
+        /// <param name="height">Block's height.</param>
+        /// <param name="block">Block for which reward amount is checked.</param>
+        /// <exception cref="ConsensusErrors.BadCoinbaseAmount">Thrown if coinbase transaction output value is larger than expected.</exception>
+        public abstract void CheckBlockReward(RuleContext context, Money fees, int height, Block block);
+
+        /// <summary>
+        /// Network specific proof of work reward amount for the block at provided height.
+        /// </summary>
+        /// <param name="height">Height of the block that we're calculating the reward for.</param>
+        /// <returns>Reward amount.</returns>
+        public abstract Money GetProofOfWorkReward(int height);
+
+        /// <summary>
+        /// Checks the maturity of UTXOs.
+        /// </summary>
+        /// <param name="coins">UTXOs to check the maturity of.</param>
+        /// <param name="spendHeight">Height at which coins are attempted to be spent.</param>
+        /// <exception cref="ConsensusErrors.BadTransactionPrematureCoinbaseSpending">Thrown if transaction tries to spend coins that are not mature.</exception>
+        public abstract void OnCheckMaturity(UnspentOutputs coins, int spendHeight);
+
+        /// <summary>
+        /// Network specific initialization logic.
+        /// </summary>
+        public abstract void OnInitialize();
+
+        /// <summary>
+        /// Network specific updates to the context's UTXO set.
+        /// </summary>
+        /// <param name="context">Context that contains variety of information regarding blocks validation and execution.</param>
+        /// <param name="transaction">Transaction which outputs will be added to the context's <see cref="UnspentOutputSet"/> and which inputs will be removed from it.</param>
+        public abstract void OnUpdateCoinView(RuleContext context, Transaction transaction);
+
+        /// <inheritdoc />
+        protected async Task OnRunAsync(RuleContext context)
+        {
+            this.Logger.LogTrace("()");
+
+            Block block = context.BlockValidationContext.Block;
+            ChainedHeader index = context.BlockValidationContext.ChainedHeader;
+            DeploymentFlags flags = context.Flags;
+            UnspentOutputSet view = context.Set;
+
+            this.Parent.PerformanceCounter.AddProcessedBlocks(1);
+
+            long sigOpsCost = 0;
+            Money fees = Money.Zero;
+            var checkInputs = new List<Task<bool>>();
+            for (int txIndex = 0; txIndex < block.Transactions.Count; txIndex++)
+            {
+                this.Parent.PerformanceCounter.AddProcessedTransactions(1);
+                Transaction tx = block.Transactions[txIndex];
+                if (!context.SkipValidation)
+                {
+                    // TODO: Simplify this condition.
+                    if (!tx.IsCoinBase && (!context.IsPoS || (context.IsPoS && !tx.IsCoinStake)))
```

Comment:
If we already have two implementations this logic should be an abstract in the core class (so we dont have to mention POS in a POW network).

Reply:
Ah yes @dangershony well spotted! I didn't want to make any logic changes in this PR but this is definitely something that we can look at in a PR after this one 👍